### PR TITLE
8331467: FileSystems.getDefault fails with ClassNotFoundException if custom default provider is in run-time image

### PR DESCRIPTION
**Pull request by liyazzi**
For two cases:

1. When the ImageReaderFactory was loaded by local jdk,that means the ImageReaderFactory was loaded by boot class loader,then init the `Path BOOT_MODULES_JIMAGE` by using `sun.nio.fs.DefaultFileSystemProvider` which is obtained through reflection,due to it is in jdk internal.
2. When loaded by a target jdk, such as jdk8 runtime, then use the Java 8 compatible APIs: `FileSystems.getDefault()` to init the `BOOT_MODULES_JIMAGE` field.
Then we can avoid the circular dependencies in class loading caused by loading the defaultSystemProvider.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331467](https://bugs.openjdk.org/browse/JDK-8331467): FileSystems.getDefault fails with ClassNotFoundException if custom default provider is in run-time image (**Bug** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22628/head:pull/22628` \
`$ git checkout pull/22628`

Update a local copy of the PR: \
`$ git checkout pull/22628` \
`$ git pull https://git.openjdk.org/jdk.git pull/22628/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22628`

View PR using the GUI difftool: \
`$ git pr show -t 22628`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22628.diff">https://git.openjdk.org/jdk/pull/22628.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22628#issuecomment-2525040560)
</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
This is the 3rd PR created for this issue, would it be possible to say if you are working with jyxzwd on this issue?

**Comment by liyazzi** [comment]
> This is the 3rd PR created for this issue, would it be possible to say if you are working with jyxzwd on this issue?

Yeah, There were something wrong with his OCA, so I re-issued this PR

**Comment by liyazzi** [comment]
All of the `test (hs/tier1 compiler not-xcomp)` failed. Does it  matter?

**Comment by liyazzi** [comment]
> All of the `test (hs/tier1 compiler not-xcomp)` failed. Does it matter?

@AlanBateman Is the failure of `test (hs/tier1 compiler not-xcomp)` related to this issue? I should correct the corresponding test, right?

**Comment by AlanBateman** [comment]
The update to ImageReaderFactory is what I suggested in the other PR so that part is okay.

The proposed test is very different to what I would have expected, meaning I was surprised to see it setting BOOT_MODULES_JIMAGE and needing to run with --add-exports.  It can be massively simplified with `@build` tag to compile the test provider, using ToolProvider to run jlink with `--add-options` to set the system provider to override the default provider, e.g.
```
        String mlib = Path.of(System.getProperty("test.classes"), "modules").toString();
        String imageDir = "myimage";
        ToolProvider jlink = ToolProvider.findFirst("jlink").orElseThrow();
        int res = jlink.run(System.out, System.err,
                "--module-path", mlib,
                "--add-modules", "customfs",
                "--add-options", "-Djava.nio.file.spi.DefaultFileSystemProvider=customfs.CustomFileSystemProvider",
                "--output", imageDir);
        if (res != 0) {
            throw new RuntimeException("jlink failed");
        }

        String launcher = Path.of(imageDir, "bin", "java").toString();
        String[] cmd = { launcher, "--list-modules" };
        var pb = new ProcessBuilder(cmd);
        OutputAnalyzer outputAnalyzer = ProcessTools
                .executeProcess(pb)
                .outputTo(System.err)
                .errorTo(System.err);
        outputAnalyzer.shouldHaveExitValue(0);
```
Even better would if we updated test/jdk/java/nio/file/spi/SetDefaultProvider.java to add this as a test case. Right now, this test exercises several different configurations and this issue is really just another configuration.

**Comment by liyazzi** [comment]
In [macos-x64 / test (jdk/tier1 part 1)](https://github.com/liyazzi/jdk/actions/runs/12252749523/job/34182811710),`jdk/internal/vm/Continuation/Fuzz.java#preserve-fp]: Fuzz tests for jdk.internal.vm.Continuation` failed, Does it matter?
The log is going this:
```
 Compilation status for: int Fuzz.com_int(int,int)
isMethodCompiled: false isMethodCompilable: true isMethodQueuedForCompilation: false getMethodCompilationLevel: 0
----------System.err:(28/1898)----------
java.lang.AssertionError: Failed to compile int Fuzz.com_int(int,int) in 20000ms
	at Fuzz.waitForCompilation(Fuzz.java:324)
	at Fuzz.compile(Fuzz.java:358)
	at Fuzz.testTrace(Fuzz.java:215)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:186)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:215)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:215)
	at java.base/java.nio.file.FileChannelLinesSpliterator.forEachRemaining(FileChannelLinesSpliterator.java:130)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:153)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:176)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:636)
	at Fuzz.testStream(Fuzz.java:205)
	at Fuzz.testFile(Fuzz.java:190)
	at Fuzz.runTests(Fuzz.java:112)
	at Fuzz.main(Fuzz.java:106)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
	at java.base/java.lang.Thread.run(Thread.java:1447)

JavaTest Message: Test threw exception: java.lang.AssertionError: Failed to compile int Fuzz.com_int(int,int) in 20000ms
JavaTest Message: shutting down test

STATUS:Failed.`main' threw exception: java.lang.AssertionError: Failed to compile int Fuzz.com_int(int,int) in 20000ms
```

**Comment by liyazzi** [comment]
> In [macos-x64 / test (jdk/tier1 part 1)](https://github.com/liyazzi/jdk/actions/runs/12252749523/job/34182811710),`jdk/internal/vm/Continuation/Fuzz.java#preserve-fp]: Fuzz tests for jdk.internal.vm.Continuation` failed, Does it matter? The log is going this:
> 
> ```
>  Compilation status for: int Fuzz.com_int(int,int)
> isMethodCompiled: false isMethodCompilable: true isMethodQueuedForCompilation: false getMethodCompilationLevel: 0
> ----------System.err:(28/1898)----------
> java.lang.AssertionError: Failed to compile int Fuzz.com_int(int,int) in 20000ms
> 	at Fuzz.waitForCompilation(Fuzz.java:324)
> 	at Fuzz.compile(Fuzz.java:358)
> 	at Fuzz.testTrace(Fuzz.java:215)
> 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:186)
> 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:215)
> 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:197)
> 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:215)
> 	at java.base/java.nio.file.FileChannelLinesSpliterator.forEachRemaining(FileChannelLinesSpliterator.java:130)
> 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
> 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
> 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:153)
> 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:176)
> 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
> 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:636)
> 	at Fuzz.testStream(Fuzz.java:205)
> 	at Fuzz.testFile(Fuzz.java:190)
> 	at Fuzz.runTests(Fuzz.java:112)
> 	at Fuzz.main(Fuzz.java:106)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
> 	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
> 	at java.base/java.lang.Thread.run(Thread.java:1447)
> 
> JavaTest Message: Test threw exception: java.lang.AssertionError: Failed to compile int Fuzz.com_int(int,int) in 20000ms
> JavaTest Message: shutting down test
> 
> STATUS:Failed.`main' threw exception: java.lang.AssertionError: Failed to compile int Fuzz.com_int(int,int) in 20000ms
> ```

I pulled the latest changes from the main branch and then the tests all ran successfully.

**Comment by AlanBateman** [comment]
The test change means this test ends up with two copies of the test provider. I'll create an issue in JBS to refactor this test to allow it be used for all cases (class path, module path, linked into run-time image). That would make this much simpler.

The bigger question is whether specification of FileSystems.getDefault should be revised to specify how it should work when the default file system provider is deployed as a module (which would cover the case where there the provider has been linked into the a run-time image). The reasons the tests work is because the provider module exports the package with the provider class but it should be possible to fully encapsulate it. I'll try to do some think on what the right thing to do there is.

**Comment by liyazzi** [comment]
> The bigger question is whether specification of FileSystems.getDefault should be revised to specify how it should work when the default file system provider is deployed as a module (which would cover the case where there the provider has been linked into the a run-time image). The reasons the tests work is because the provider module exports the package with the provider class but it should be possible to fully encapsulate it. I'll try to do some think on what the right thing to do there is.

For the specification of FileSystems.getDefault,In my opinion, I think there would be 2 things need to be added.

**A**: add instructions for paying attention to circular dependencies when loading the default file system provider. The changes implemented by this PR to ImageFactory have solved the circular dependency problem in the case where the default file system provider is deployed as a module.

**B**: specify what we should do to keep full encapsulation when loading the default file system provider(maybe not one,but a list).The current performance of FileSystems.getDefault is that:

1. If the default file system provider is on classpath,then it will be in the Unnamed-Module which exports the package that contains the default file system provider.And it is okay.
2. If it is on module path,which means the default file system provider is in a Named-Module(this includes the case of linked into run-time image),unless we export the package with the provider class,it would throw IllegalAccessException: java.base -> customfs module.

Because in current specification of FileSystems.getDefault it is going this:

> Each class is loaded, **using the system class loader**, and instantiated by invoking a one argument constructor whose formal parameter type is FileSystemProvider.

It simply uses the `ClassLoader.getSystemClassLoader()` then leading to the performance mentioned above.

For the **B** point, my opinion is that: not just using the system class loader,but we could load provider class like java SPI loads the Service-Provider.Specifically, the code is as follows(change the FileSystems$DefaultFileSystemHolder.getDefaultProvider):

```java
        // returns default provider
        private static FileSystemProvider getDefaultProvider() {
            // start with the platform's default file system provider
            FileSystemProvider provider = DefaultFileSystemProvider.instance();

            // if the property java.nio.file.spi.DefaultFileSystemProvider is
            // set then its value is the name of the default provider (or a list)
            String prop = "java.nio.file.spi.DefaultFileSystemProvider";
            String propValue = System.getProperty(prop);
            if (propValue != null) {
                boolean moduleSpecified;
                Class<?> c;
                Constructor<?> ctor;
                // the format of cn should be like this: [<module-name>/]<fully-qualified-class-name>
                for (String cn: propValue.split(",")) {
                    try {
                        moduleSpecified = cn.contains("/");
                        if(!moduleSpecified) {
                            c = Class
                                    .forName(cn, true, ClassLoader.getSystemClassLoader());
                            ctor = c
                                    .getDeclaredConstructor(FileSystemProvider.class);
                        }else {
                            // get the module name
                            int spiltIndex = cn.indexOf('/');
                            String moduleName = cn.substring(0, spiltIndex);
                            // get the class name
                            cn = cn.substring(spiltIndex +1);
                            Module module = ModuleLayer.boot().findModule(moduleName).orElseThrow(()->new Error("module not found: " + moduleName));
                            c = Class.forName(module,cn);
                            ctor = c.getDeclaredConstructor(FileSystemProvider.class);
                            // get the access to the provider class without exports package which contains default provider class
                            ctor.setAccessible(true);
                        }

                        provider = (FileSystemProvider)ctor.newInstance(provider);

                        // must be "file"
                        if (!provider.getScheme().equals("file"))
                            throw new Error("Default provider must use scheme 'file'");

                    } catch (Exception x) {
                        throw new Error(x);
                    }
                }
            }
            return provider;
        }
```

In the above code, it changes 2 things:

1. Change the format of the value of the `java.nio.file.spi.DefaultFileSystemProvider` to `[<module-name>/]<fully-qualified-class-name> ...` e.g. `-Djava.nio.file.spi.DefaultFileSystemProvider=module1/foo.cutomfsprovider1,bar.cutomfsprovider2`
2. If `<module-name>` is not specified, then load provider as usual; else load provider from its module without provider exporting its package.

I test with these cases:

test resource: two modular jar

1. **customfs** module: customfs.jar (contains the custom provider class: `foo.CustomFileSystemProvider`; export nothing)
2. **main** module: main.jar (contains main class: `bar.main` to run)

- class-path case: `java -cp customfs.jar:main.jar -Djava.nio.file.spi.DefaultFileSystemProvider=foo.CustomFileSystemProvider bar.Main` success

- module-path case: `java -p customfs.jar:main.jar -Djava.nio.file.spi.DefaultFileSystemProvider=customfs/foo.CustomFileSystemProvider --add-modules customfs -m main/bar.Main` diff: **foo.CustomFileSystemProvider -> customfs/er=foo.CustomFileSystemProvider** and success

- linked-into image case(contains **customfs**, **main**, **java.base** modules): `image/bin/java -Djava.nio.file.spi.DefaultFileSystemProvider=customfs/foo.CustomFileSystemProvider -m main/bar.Main` 
**failed: java.lang.Error: module not found: customfs**. I found that even in image, only if add `--add-modules customfs` option,would jvm add the **customfs** module to the `ModuleLayer.boot()`.
So: `image/bin/java -Djava.nio.file.spi.DefaultFileSystemProvider=customfs/foo.CustomFileSystemProvider --add-modules customfs -m main/bar.Main`  success

**Comment by AlanBateman** [comment]
@liyazzi  JDK-8346463 has been integrated so I think we have a better base to add more tests. It includes a `@Disabled` test that run jlink to create a run-time image that includes the custom default provider so I think it can be used here.

**Comment by liyazzi** [comment]
In addition to 'testFspInRuntimeImage', we should add more tests , right? e.g. 

1. The file system provider linked into run-time image and the application in exploded module on module path.
2. The file system provider linked into run-time image and the application in modular JAR on module path.
3. Or both of the file system provider and the application are in the same run-time image with two module.

**Review by AlanBateman — COMMENTED** [review]
I'm happy with the current changes in the PR. The copyright header date in ImageReaderFactory can be updated to 2024, and the import of Disabled can be removed from the SetDefaultProvider test.

To your question about tests then it's possible. The test that has been enabled in this change is the main test we will to ensure this configuration works. So I think we can leave further tests in this area to further PRs.

**Comment by liyazzi** [comment]
> I'm happy with the current changes in the PR. The copyright header date in ImageReaderFactory can be updated to 2024, and the import of Disabled can be removed from the SetDefaultProvider test.
> 
> To your question about tests then it's possible. The test that has been enabled in this change is the main test we will to ensure this configuration works. So I think we can leave further tests in this area to further PRs.

Okay, thank you for your review!

File: src/java.base/share/classes/jdk/internal/jimage/ImageReaderFactory.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2015, 2016, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/jimage/ImageReaderFactory.java:2]: It should be changed to "2015, 2024,"

**Review by AlanBateman — APPROVED** [review]

**Review by jaikiran — APPROVED** [review]
The changes in this PR which are backed by the discussions in https://github.com/openjdk/jdk/pull/21997 look fine to me.

**Comment by AlanBateman** [comment]
> The changes in this PR which are backed by the discussions in #21997 look fine to me.

Thanks for putting more eyes on this.

**Pull request closed without merging**